rp235x-hal/src/hstx.rs

```diff
@@ -0,0 +1,324 @@
+//! let mut peripherals = hal::pac::Peripherals::take().unwrap();
+//! let sio = Sio::new(peripherals.SIO);
+//! let pins = hal::gpio::Pins::new(
+//!     peripherals.IO_BANK0,
+//!     peripherals.PADS_BANK0,
+//!     sio.gpio_bank0,
+//!     &mut peripherals.RESETS,
+//! );
+//!
+//! let bit0p = pins.gpio12.into_function();
+//! let bit3p = pins.gpio15.into_function();
+//!
+//! let hstx_pins = HstxPins::new().add_bit0_pin(bit0p).add_bit3_pin(bit3p);
+//!
+//! let mut hstx = Hstx::new(peripherals.HSTX_CTRL, peripherals.HSTX_FIFO, hstx_pins,
+//!                          &mut peripherals.RESETS);
+//! hstx.configure_clk(5, 0);
+//! hstx.configure_shifter(2, 5);
+//! hstx.configure_bit0(HstxBitConfig::Clk { inv: false });
+//! hstx.configure_bit3(HstxBitConfig::Shift { sel_p:0, sel_n: 1, inv: false });
+//!
+//! let hstx = hstx.enable();
+//!
+//! loop {
+//!     hstx.fifo_write(0xaabbccdd);
+//! }
+//! ```
+
+use crate::{
+    dma::{EndlessWriteTarget, WriteTarget},
+    pac::{dma::ch::ch_ctrl_trig::TREQ_SEL_A, HSTX_CTRL, HSTX_FIFO, RESETS},
+    resets::SubsystemReset,
+    typelevel::{Is, OptionTSome, Sealed},
+};
+use core::marker::PhantomData;
+
+mod pins;
+pub use pins::*;
+/// State of HSTX
+pub trait State: Sealed {}
+
+/// HSTX is disabled, FIFO is not being drained
+pub struct Disabled {
+    __private: (),
+}
+
+/// HSTX is enabled, data from FIFO is being consumed
+pub struct Enabled {
+    __private: (),
+}
+
+impl State for Disabled {}
+impl Sealed for Disabled {}
+impl State for Enabled {}
+impl Sealed for Enabled {}
+
+/// HSTX peripheral
+pub struct Hstx<S: State, P: ValidHstxPinout> {
+    ctrl: HSTX_CTRL,
+    fifo: HSTX_FIFO,
+    pins: P,
+    state: PhantomData<S>,
+}
+
+impl<P: ValidHstxPinout> Hstx<Disabled, P> {
+    /// Instantiate and reset HSTX peripheral
+    pub fn new(
+        ctrl: HSTX_CTRL,
+        fifo: HSTX_FIFO,
+        pins: P,
+        resets: &mut RESETS,
+    ) -> Hstx<Disabled, P> {
+        ctrl.reset_bring_down(resets);
+        ctrl.reset_bring_up(resets);
+        Hstx {
+            ctrl,
+            fifo,
+            pins,
+            state: PhantomData,
+        }
+    }
+
+    /// Free HSTX peripheral and get registers and pins back
+    pub fn free(self) -> (HSTX_CTRL, HSTX_FIFO, P) {
+        (self.ctrl, self.fifo, self.pins)
+    }
+
+    /// Configure clock generator
+    pub fn configure_clk(&self, div: u8, phase: u8) {
+        if phase >= 2 * div {
+            panic!("phase should be strictly less than double the div");
+        }
+
+        self.ctrl
+            .csr()
+            .modify(|_, w| unsafe { w.clkdiv().bits(div).clkphase().bits(phase) });
+    }
+
+    /// Configure output shifter
+    pub fn configure_shifter(&self, shift: u8, n_shifts: u8) {
+        self.ctrl
+            .csr()
+            .modify(|_, w| unsafe { w.shift().bits(shift).n_shifts().bits(n_shifts) });
+    }
+
+    /// Configure expand shifter
+    pub fn configure_expand_shifter(
+        &self,
```

Comment:
These methods are changing shared mutable state in the hardware and so should require &mut I think.